src/analysis/normed_space/star/multiplier.lean

```diff
@@ -0,0 +1,450 @@
+
+lemma add_fst (a b : 𝓜(𝕜, A)) : (a + b).fst = a.fst + b.fst := rfl
+lemma add_snd (a b : 𝓜(𝕜, A)) : (a + b).snd = a.snd + b.snd := rfl
+lemma zero_fst : (0 : 𝓜(𝕜, A)).fst = 0 := rfl
+lemma zero_snd : (0 : 𝓜(𝕜, A)).snd = 0 := rfl
+lemma neg_fst (a : 𝓜(𝕜, A)) : (-a).fst = -a.fst := rfl
+lemma neg_snd (a : 𝓜(𝕜, A)) : (-a).snd = -a.snd := rfl
+lemma sub_fst (a b : 𝓜(𝕜, A)) : (a - b).fst = a.fst - b.fst := rfl
+lemma sub_snd (a b : 𝓜(𝕜, A)) : (a - b).snd = a.snd - b.snd := rfl
+lemma one_fst : (1 : 𝓜(𝕜, A)).fst = 1 := rfl
+lemma one_snd : (1 : 𝓜(𝕜, A)).snd = 1 := rfl
+@[simp] lemma mul_fst (a b : 𝓜(𝕜, A)) : (a * b).fst = a.fst * b.fst := rfl
+@[simp] lemma mul_snd (a b : 𝓜(𝕜, A)) : (a * b).snd = b.snd * a.snd := rfl
+lemma nat_cast_fst (n : ℕ) : (n : 𝓜(𝕜 , A)).fst = n := rfl
+lemma nat_cast_snd (n : ℕ) : (n : 𝓜(𝕜 , A)).snd = n := rfl
+lemma int_cast_fst (n : ℤ) : (n : 𝓜(𝕜 , A)).fst = n := rfl
+lemma int_cast_snd (n : ℤ) : (n : 𝓜(𝕜 , A)).snd = n := rfl
+lemma pow_fst (n : ℕ) (a : 𝓜(𝕜, A)) : (a ^ n).fst = a.fst ^ n := rfl
+lemma pow_snd (n : ℕ) (a : 𝓜(𝕜, A)) : (a ^ n).snd = a.snd ^ n := rfl
+
+/-- The natural injection from `double_centralizer.to_prod` except the second coordinate inherits
+`mul_opposite.op`. The ring structure on `𝓜(𝕜, A)` is the pullback under this map. -/
+def to_prod_mul_opposite : 𝓜(𝕜, A) → (A →L[𝕜] A) × (A →L[𝕜] A)ᵐᵒᵖ :=
+λ a, (a.fst, mul_opposite.op a.snd)
+
+lemma to_prod_mul_opposite_injective :
+  function.injective (to_prod_mul_opposite : 𝓜(𝕜, A) → (A →L[𝕜] A) × (A →L[𝕜] A)ᵐᵒᵖ) :=
+λ a b h, let h' := prod.ext_iff.mp h in ext _ _ $ prod.ext h'.1 $ mul_opposite.op_injective h'.2
+
+/-- The ring structure is inherited as the pullback under the injective map
+`double_centralizer.to_prod_mop : 𝓜(𝕜, A) → (A →L[𝕜] A) × (A →L[𝕜] A)ᵐᵒᵖ` -/
+instance : ring 𝓜(𝕜, A) :=
+to_prod_mul_opposite_injective.ring _
+  rfl rfl (λ _ _, rfl) (λ _ _, rfl) (λ _, rfl) (λ _ _, rfl)
+  (λ x n, prod.ext rfl $ mul_opposite.op_smul _ _)
+  (λ x n, prod.ext rfl $ mul_opposite.op_smul _ _)
+  (λ x n, prod.ext rfl $ mul_opposite.op_pow _ _)
+  (λ _, rfl) (λ _, rfl)
+
+/-- The canonical map `double_centralizer.to_prod` as an additive group homomorphism. -/
+def to_prod_hom : 𝓜(𝕜, A) →+ (A →L[𝕜] A) × (A →L[𝕜] A) :=
+{ to_fun := to_prod,
+  map_zero' := rfl,
+  map_add' := λ x y, rfl }
+
+/-- The module structure is inherited as the pullback under the additive group monomorphism
+`double_centralizer.to_prod : 𝓜(𝕜, A) →+ (A →L[𝕜] A) × (A →L[𝕜] A)` -/
+instance : module 𝕜 𝓜(𝕜, A) :=
+function.injective.module 𝕜 to_prod_hom ext (λ x y, rfl)
+
+instance : algebra 𝕜 𝓜(𝕜, A) :=
+algebra.of_module
+  (λ k a b, by {ext; simp only [mul_fst, smul_fst, mul_snd, smul_snd, coe_smul',pi.smul_apply,
+    continuous_linear_map.coe_mul, function.comp_app, continuous_linear_map.map_smul]})
+  (λ k a b, by {ext; simp only [mul_fst, smul_fst, mul_snd, smul_snd, algebra.mul_smul_comm,
+    coe_smul', continuous_linear_map.coe_mul, pi.smul_apply, function.comp_app]})
```

Comment:
Can you add lemmas about `fst (algebra_map _ 𝓜(𝕜, A) x)` and `snd (algebra_map _ 𝓜(𝕜, A) k)`?

Reply:
I decided to explicitly provide the algebra instance instead of using `algebra.to_module` so that we could get nice definitional unfolding with `to_prod`, and I added the relevant lemmas about `algebra_map`.